File: git_diff_exporter.py
#!/usr/bin/env python3

import argparse
import os
import shutil
import subprocess
from datetime import datetime

parser = argparse.ArgumentParser(description='Export diff files between two commits.')
parser.add_argument('commit1', help='The first commit hash')
parser.add_argument('commit2', help='The second commit hash')
parser.add_argument('-o', '--output', help='Output path', default='..')
parser.add_argument('-f', '--format', help='Output name format (ex. OUTPUT-%Y%m%d)', default=None)
args = parser.parse_args()

if not args.commit1 or not args.commit2:
    print("Please provide two commit hashes")
    exit(1)

diff_files_output = subprocess.check_output(['git', 'diff', '--name-only', args.commit1, args.commit2]).decode('utf-8').strip()
diff_files = diff_files_output.split('\n')

if not diff_files_output:
    print("No files to export")
    exit(0)

output_name = args.format.format() if args.format else f"diff-{args.commit1[0:7]}-{args.commit2[0:7]}"
output_path = os.path.abspath(args.output)

if not os.path.exists(output_path):
    os.makedirs(output_path)

for file in diff_files:
    source_path = os.path.join(os.getcwd(), file)
    target_path = os.path.join(output_path, output_name, file)

    if os.path.exists(source_path):
        os.makedirs(os.path.dirname(target_path), exist_ok=True)
        shutil.copy2(source_path, target_path)
        print(f"Copied: {source_path} -> {target_path}")

print(f"Exported diff to {os.path.join(output_path, output_name)}")

File: git_diff_exporter.rb
#!/usr/bin/env ruby

require 'optparse'
require 'fileutils'

options = {
  output: nil,
  format: nil
}

OptionParser.new do |opts|
  opts.banner = "Usage: git-export-diff.rb [options] commit1 commit2"

  opts.on("-oOUTPUT", "--output=OUTPUT", "Output path") do |output|
    options[:output] = output
  end

  opts.on("-fFORMAT", "--format=FORMAT", "Output name format (ex. OUTPUT-%Y%m%d)") do |format|
    options[:format] = format
  end
end.parse!

if ARGV.size != 2
  puts "Please provide two commit hashes"
  exit 1
end

commit1, commit2 = ARGV

diff_files = `git diff --name-only #{commit1} #{commit2}`.strip.split("\n")

if diff_files.empty?
  puts "No files to export"
  exit 0
end

output_name = options[:format] ? Time.now.strftime(options[:format]) : "diff-#{commit1[0, 7]}-#{commit2[0, 7]}"
output_path = File.expand_path(options[:output] || "..")

FileUtils.mkdir_p(output_path) unless Dir.exist?(output_path)

diff_files.each do |file|
  source_path = File.join(Dir.pwd, file)
  target_path = File.join(output_path, output_name, file)

  if File.exist?(source_path)
    FileUtils.mkdir_p(File.dirname(target_path))
    FileUtils.cp(source_path, target_path)
    puts "Copied: #{source_path} -> #{target_path}"
  end
end

puts "Exported diff to #{File.join(output_path, output_name)}"